src/Grid/grid.jl

```diff
@@ -68,12 +68,133 @@ struct VertexIndex <: BoundaryIndex
+end
+
+Neighbor(info::T) where T <: BoundaryIndex = Neighbor([info])
+Base.zero(::Type{Neighbor}) = Neighbor(BoundaryIndex[])
+Base.length(n::Neighbor) = length(n.neighbor_info)
+Base.getindex(n::Neighbor,i) = getindex(n.neighbor_info,i)
+Base.firstindex(n::Neighbor) = 1
+Base.lastindex(n::Neighbor) = length(n.neighbor_info)
+Base.:(==)(n1::Neighbor, n2::Neighbor) = n1.neighbor_info == n2.neighbor_info
+
+function Base.:+(n1::Neighbor, n2::Neighbor)
+    neighbor_info = [n1.neighbor_info; n2.neighbor_info]
+    return Neighbor(neighbor_info)
+end
+
+function Base.show(io::IO, ::MIME"text/plain", n::Neighbor)
+    if length(n) == 0
+        println(io, "No Neighbor")
+    elseif length(n) == 1
+        println(io, "$(n.neighbor_info[1])")
+    else
+        println(io, "$(n.neighbor_info...)")
+    end
+end
+
+face_npoints(::Cell{2,N,M}) where {N,M} = 2
+face_npoints(::Cell{3,4,1}) = 4 #not sure how to handle embedded cells e.g. Quadrilateral3D
+edge_npoints(::Cell{3,4,1}) = 2 #not sure how to handle embedded cells e.g. Quadrilateral3D
+face_npoints(::Cell{3,N,6}) where N = 4
+face_npoints(::Cell{3,N,4}) where N = 3
+edge_npoints(::Cell{3,N,M}) where {N,M} = 2
+
+getdim(::Cell{dim,N,M}) where {dim,N,M} = dim
+
+abstract type AbstractTopology end
+
+struct GridTopology <: AbstractTopology
+    face_neighbor::SparseMatrixCSC{Neighbor,Int}
+    corner_neighbor::SparseMatrixCSC{Neighbor,Int}
+    edge_neighbor::SparseMatrixCSC{Neighbor,Int}
+end
+
+function GridTopology()
+    face_neighbor = spzeros(Neighbor,0,0)
+    corner_neighbor = spzeros(Neighbor,0,0)
+    edge_neighbor = spzeros(Neighbor,0,0)
+    return GridTopology(face_neighbor,corner_neighbor,edge_neighbor)
+end
+
+function GridTopology(cells::Vector{C}) where C <: AbstractCell
+    I_face = Int[]; J_face = Int[]; V_face = Neighbor[]
+    I_edge = Int[]; J_edge = Int[]; V_edge = Neighbor[]
+    I_corner = Int[]; J_corner = Int[]; V_corner = Neighbor[]
+    cell_node_table = getproperty.(cells,:nodes)
+    for (cellid,cell) in enumerate(cells)
+        neighbors = findall.(x->x ∈ cell.nodes,cell_node_table) 
+        for (neighborid,neighbor) in enumerate(neighbors)
+            neighbor_cell = cells[neighborid]
+            if length(neighbor) == 0
+                #not a neighbor
+                continue
+            elseif length(neighbor) == 1
+                #corner neighbor
+                _corner_neighbor!(V_corner,I_corner,J_corner,cellid,cell,neighbor,neighborid,neighbor_cell)
+            elseif getdim(cell) > 2 && length(neighbor) == edge_npoints(cell) 
+                #edge neighbor
+                _edge_neighbor!(V_edge,I_edge,J_edge,cellid,cell,neighbor,neighborid,neighbor_cell)
+            elseif getdim(cell) > 1 && length(neighbor) == face_npoints(cell)
+                #face neighbor
+                _face_neighbor!(V_face,I_face,J_face,cellid,cell,neighbor,neighborid,neighbor_cell)
+            else
+                continue
+            end 
+        end
+    end
```

Comment:
This construction is quite expensive ( O(#cells * #nodes), which is basically quadratic in the number of cells ).

My alternative construction would be to derive the topology as follows.

1. vertex -> cell ("invert" cell -> vertex by looping over the vertices)
1. now you can derive cell -> cell via the concatenation cell -> vertex -> cell
1. now the face neighborhood can be found by cell -> cell and comparing the unique face identifiers
1. same for the edge neighborhood

Reply:
should be resolved